### ISSUE TITLE
[react-hooks/exhaustive-deps] generic parameter is reported as depended upon (new scope analyzer bug?)

### ISSUE THREAD

**Issue by AviVahl** [typescript-eslint/typescript-eslint#2479]
This issue was observed in a very large repo with a lot of React components.
Some of those components are generic ones, and create functions that use `T` in their parameter list.
Since upgrading to typescript-eslint@4.0.1, `react-hooks/exhaustive-deps` started reporting `T` as a missing dependency.

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

https://github.com/wixplosives/sample-monorepo/blob/master/.eslintrc
https://github.com/wixplosives/sample-monorepo/blob/master/tsconfig.json

The original code is closed source and contains a lot of unrelated stuff. Here's a smaller reproduction
```TS
export function Comp<T>(): JSX.Element | null {
  const myFn = React.useCallback((v: T) => {
    return v;
  }, []); // <----- lint error occurs here

  // ...

  return null;
}
```

**Expected Result**

No `react-hooks/exhaustive-deps` error.

**Actual Result**

`React Hook React.useCallback has a missing dependency: 'T'. Either include it or remove the dependency array.`

The rule suddenly offers the `T` from the parameter list as a dependency. 

**Additional Info**

Did not happen in v3.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.0.1` |
| `@typescript-eslint/parser`        | `4.0.1` |
| `TypeScript`                       | `4.0.2` |
| `ESLint`                           | `7.8.1` |
| `node`                             | `14.9.0` |

**Comment by bradzacher** [comment]
Duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/2467

This needs to be fixed in the rule.

**Comment by AviVahl** [comment]
Thanks for the quick response.
Found https://github.com/facebook/react/pull/19751
